File: week02/이분탐색/1920.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline
N = int(input()) 
A_list = list(map(int, input().split())) 
M = int(input()) 
M_list = list(map(int, input().split())) 

A_list.sort() 

# def binary_search(target, start, end):
#   if start > end:
#     return 0
  
#   mid = (start + end) // 2

#   if A_list[mid] == target:
#     return 1
#   elif A_list[mid] > target:
#     end = mid - 1
#   else:
#     start = mid + 1

#   return binary_search(target, start, end)

def binary_search(target, data):
  start = 0
  end = len(data) - 1

  while start <= end:
    mid = (start + end) // 2 
    if data[mid] == target:
      return 1
    elif data[mid] > target:
      end = mid - 1
    else:
      start = mid + 1
  return 0


for elems in M_list:
  print(binary_search(elems, A_list))

File: week02_JS/이분탐색/1920.js
// M개의 정수가 N개의 정수 리스트 안에 있는지 확인

const fs = require('fs');
const input = fs.readFileSync('input.txt').toString().trim().split('\n');

let N = parseInt(input[0]);
let A_list = input[1].split(' ').map(Number)
let M = parseInt(input[2])
let M_list = input[3].split(' ').map(Number)

let result = []

// N개의 정수 정렬
A_list.sort((a, b) => a - b)
console.log(A_list)
// 이분 탐색
function binary_search(arr, target) {
  let start = 0
  let end = arr.length - 1

  while (start <= end) {
    let mid = Math.ceil((start + end) / 2)
    if (arr[mid] == target) {
      result.push(1)
      return
    } else if (arr[mid] > target) {
      end = mid - 1
    } else {
      start = mid + 1
    }
  }
  result.push(0)
  return
}

M_list.forEach(elems => binary_search(A_list, elems))
console.log(result.join('\n'))